theta_dup/include/theta_sketch_dup.h

```diff
@@ -0,0 +1,2209 @@
+                                                        uint16_t seed_hash) {
+  uint64_t theta = theta_sketch_dup_alloc<A>::MAX_THETA;
+  uint32_t num_keys = 0;
+
+  const bool is_empty =
+      flags_byte & (1 << theta_sketch_dup_alloc<A>::flags::IS_EMPTY);
+  if (!is_empty) {
+    if (preamble_longs == 1) {
+      num_keys = 1;
+    } else {
+      is.read((char*)&num_keys, sizeof(num_keys));
+      uint32_t unused32;
+      is.read((char*)&unused32, sizeof(unused32));
+      if (preamble_longs > 2) {
+        is.read((char*)&theta, sizeof(theta));
+      }
+    }
+  }
+  vector_u64<A> keys(num_keys);
+  if (!is_empty)
+    is.read((char*)keys.data(),
+            sizeof(std::pair<uint64_t, uint64_t>) * keys.size());
+
+  const bool is_ordered =
+      flags_byte & (1 << theta_sketch_dup_alloc<A>::flags::IS_ORDERED);
+  if (!is.good()) throw std::runtime_error("error reading from std::istream");
+  if (!is.good()) throw std::runtime_error("error reading from std::istream");
+  return compact_theta_sketch_dup_alloc<A>(is_empty, theta, std::move(keys),
+                                           seed_hash, is_ordered);
+}
+
+template <typename A>
+compact_theta_sketch_dup_alloc<A>
+compact_theta_sketch_dup_alloc<A>::deserialize(
+    datasketches_pb::Update_theta_sketch_dup& pb, uint64_t seed) {
+  uint8_t preamble_longs = pb.preamble_longs();
+  uint8_t serial_version = pb.serial_version();
+  uint8_t type = pb.sketch_type();
+  uint8_t flags_byte = pb.flags_byte();
+  uint16_t seed_hash = pb.seed_hash();
+  theta_sketch_dup_alloc<A>::check_sketch_type(type, SKETCH_TYPE);
+  theta_sketch_dup_alloc<A>::check_serial_version(
+      serial_version, theta_sketch_dup_alloc<A>::SERIAL_VERSION);
+  theta_sketch_dup_alloc<A>::check_seed_hash(
+      seed_hash, theta_sketch_dup_alloc<A>::get_seed_hash(seed));
+  return internal_deserialize(pb, preamble_longs, flags_byte, seed_hash);
+}
+
+template <typename A>
+compact_theta_sketch_dup_alloc<A>
+compact_theta_sketch_dup_alloc<A>::internal_deserialize(
+    datasketches_pb::Update_theta_sketch_dup& pb, uint8_t flags_byte,
+    uint8_t preamble_longs, uint16_t seed_hash) {
+  uint64_t theta = theta_sketch_dup_alloc<A>::MAX_THETA;
+  uint32_t num_keys = 0;
+
+  const bool is_empty =
+      flags_byte & (1 << theta_sketch_dup_alloc<A>::flags::IS_EMPTY);
+  if (!is_empty) {
+    if (preamble_longs == 1) {
+      num_keys = 1;
+    } else {
+      num_keys = pb.num_keys();
+      if (preamble_longs > 2) {
+        theta = pb.theta();
+      }
+    }
+  }
+  vector_u64<A> keys(num_keys);
+  if (!is_empty)
+    for (int i = 0; i < pb.keys_size(); i++)
+      keys[i] = std::make_pair(pb.keys(i).hash_val(), pb.keys(i).cnt());
+
+  const bool is_ordered =
+      flags_byte & (1 << theta_sketch_dup_alloc<A>::flags::IS_ORDERED);
+  return compact_theta_sketch_dup_alloc<A>(is_empty, theta, std::move(keys),
+                                           seed_hash, is_ordered);
+}
```

Comment:
Definition: **derived class** `compact_theta_sketch_dup_alloc`, `deserialize` (and `internal_deserialize`) using **protobuf**.